controllers/pod-mutator/controller.go

```diff
@@ -0,0 +1,70 @@
+package podmutator
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/go-logr/logr"
+)
+
+type Controller struct {
+	client.Client
+	mutators []Mutator
+	decoder  *admission.Decoder
+	log      logr.Logger
+}
+
+type Mutator interface {
+	MutatePod(ctx context.Context, pod *corev1.Pod, namespace string) error
+}
+
+// Check if the Handler interface is implemented.
+var _ admission.Handler = &Controller{}
+
+// +kubebuilder:webhook:path=/mutate-pods,mutating=true,failurePolicy=ignore,groups="",resources=pods,verbs=create,versions=v1,name=pod-mutator.storageos.com
+
+// NewController returns a new Pod mutating admission controller.
+func NewController(k8s client.Client, decoder *admission.Decoder, mutators []Mutator) *Controller {
+	return &Controller{
+		mutators: mutators,
+		Client:   k8s,
+		decoder:  decoder,
+		log:      ctrl.Log,
+	}
+}
+
+// Handle handles an admission request and mutates a pod object in the request.
+func (c *Controller) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+
+	err := c.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Get the request namespace. This is needed because sometimes the decoded
+	// pod object lacks namespace info.
+	namespace := req.AdmissionRequest.Namespace
+
+	// Create a copy of the pod to mutate.
+	copy := pod.DeepCopy()
```

Comment:
Why do you have to create a copy of the pod as opposed to just modifying the `pod` object directly and sending that back?

Reply:
This code is based on https://github.com/storageos/cluster-operator/blob/v2.3.4/internal/pkg/admission/scheduler/handler.go#L64 where we used to pass the pod object to `PatchResponse` at the very end of the function to return a response. In the latest webhook package of controller runtime, `PatchResponse` is replaced with `PatchResponseFromRaw` which accepts the raw object in the request. So, we avoid doing a deep copy here.

Reply:
Removed the `DeepCopy()`